docs/rfc/rfc-010-p2p-light-client.rst

```diff
@@ -0,0 +1,127 @@
+
+We expect to be able to leverage libp2p tools to provide some kind of service
+discovery for tendermint-based networks. This means that it will be possible
+for the p2p stack to easily identify specialized nodes, (e.g. light clients)
+thus obviating many of the design challenges with providing this feature in
+the context of the current stack.
+
+Similarly, libp2p makes it possible for project to be able back their non-Go
+light clients, without the major task of first implementing tendermint's p2p
+connection handling. We should identify if there exist users (e.g. the go IBC
+relayer, it's maintainers, and operators) who would be able to take advantage
+of p2p light client, before switching to libp2p. To our knowledge there are no
+external implementations of this p2p protocol, and it seems unlikely that a
+team would implement this directly ahead of its impending removal.
+
+User Cases
+~~~~~~~~~~
+
+This RFC makes a few assumptions about the use cases and users of light
+clients in tendermint.
+
+One of the most active use cases for light clients is in the implementation of
+the IBC relayer. Thus, we expect that providing P2P light clients might
+increase the reliability of relayers and reduce the cost of running a relayer,
+because relayer operators won't have to decide between rely on public RPC
+endpoints (unreliable) or running their own full nodes (expensive.) This
+also assumes that there are *no* other uses of the RPC in the relayer, and
+unless the relayers have the option of dropping all RPC use, it's unclear if a
+P2P light client will actually be able to successfully remove the dependency
+on the RPC system.
+
+Given that the primary relayer implementation, is hermes (rust,) it might be
+safe to deliver a version of tendermint that adds a light client rector in
+the full nodes, but that does not provide an implementation of a Go light
+client. This either means that the rust implementation would need support for
+the legacy P2P connection protocol or wait for the libp2p implementation.
+
+Client side light client (e.g. wallets, etc.) users may always want to use (a
+subset) of the RPC rather than connect to the P2P network for an ephemeral
+use.
+
+Discussion
+----------
+
+Implementation Questions
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Most of the complication in the is how to have a long lived light client node
+that *only* runs the light client reactor, as this raises a few questions:
+
+- would users specify a single P2P node to connect to when creating a light
+  client or would they also need/want to discover peers?
+
+- how do we prevent full nodes and validators from allowing their peer slots,
+  which are typically limited, from filling with light clients? If
+  light-clients aren't limited, how do we prevent light clients from consuming
+  resources on consensus nodes?
+
+- when a light client disconnects from its peers will it need to reset its
+  internal state? does this change if it connects to the same peers?
```

Comment:
What do you mean exactly by internal state? It shouldn't reset any of the blocks that it has verified. The last one will need to be used as the trusted block when it starts up again.

Reply:
I would consider it cache in the sense that persisted light blocks help to quickly verify headers that are close in height to it. The light client doesn't need to persist any more than a single verified light block and AFAIU, that's exactly how the IBC module works - it only saves the last "light block".

Reply:
That seems reasonable. I guess I was worried about a light client being (possibly) incorrect if it's getting data from a node that's providing bad data, and if it's caching data then changing the nodes you're contacting could potentially invalidate previous data. 

This might not be a valid concern.

Reply:
It's quite difficult for a node to provide invalid data that the light client believes to be valid and trustworthy. Essentially you need at least 1/3+ voting power (or more if the trust level is greater) and none of the witnesses to challenge the data. However, if that does happen then changing witnesses might help detect the divergence. If a divergence has been detected then all that cached data is now untrustworthy and the light client will have to dump it and restart with a new source of trust.

Reply:
This makes sense. I guess my mental model had been more like "if witnesses change then we have to call into question anything that we previously validated," and it seems like we only have to question things if we detect something that's invalid, which makes sense. I think the part that I'd last track of is fact that the signatures make it harder for bad data to seep into the process.

Reply:
I've updated the text with an potential answer and I think it's good to keep this in mind, but I don't think that it poses a particular problem.